### PR TITLE
fix: proper checkpoint restore weights-only vs full training state

### LINKED ISSUE #131: Properly handle difference between restoring only model weights and full training state
The `--restore_opt` argument is supposed to differentiate between restoring only the model weights, and restoring the full training session, including optimizer state (this is the crucial one) and epoch number etc. This is currently implemented as a bit of a hack: https://github.com/mllam/neural-lam/blob/43524f7065ad80c4a35d31cc8ac65a665e98138e/neural_lam/models/ar_model.py#L759-L761 that can easily start causing problems as we build on it.

A proper way to handle this in Lightning is to differentiate between instantiating the model with `load_from_checkpoint` and calling `Trainer.fit`  with `ckpt_path`. An example implementation that could be used is given here: https://github.com/joeloskarsson/neural-lam-dev/commit/e7d11c9a9b10d0165314d00f230243addd79d9fb

### PR DESCRIPTION
**Pull request by Mani212005**
## Describe your changes
Replaces the fragile `on_load_checkpoint` optimizer hack in `ARModel` with proper PyTorch
Lightning patterns for checkpoint loading. This separates two distinct behaviours:

- **Weights-only** (`--load` without `--restore_opt`): calls `ModelClass.load_from_checkpoint(path, ...)`
to instantiate the model, then `trainer.fit(model, ckpt_path=None)` — epoch/step/scheduler reset to 0.
- **Full resume** (`--load` + `--restore_opt`): uses the normal constructor, then
`trainer.fit(model, ckpt_path=path)` — restores optimizer, epoch, LR scheduler, and callbacks.

**`ar_model.py`**
- Removed `self.restore_opt = args.restore_opt` attribute.
- Removed the `on_load_checkpoint` block that manually replaced optimizer/scheduler/epoch state — kept only the key-migration logic for older checkpoints.

**`train_model.py`**
- Updated `--restore_opt` help text to say "full training state" (not just "optimizer state").
- Added assertion: `--restore_opt` requires `--load` (fails fast with a clear message).

**`test_checkpoint_restore.py`** (new)
- `test_restore_opt_without_load_raises` — assertion fires without `--load`.
- `test_weights_only_uses_load_from_checkpoint` — `load_from_checkpoint` called, `ckpt_path=None` in fit.
- `test_restore_opt_passes_ckpt_path` — normal constructor used, `ckpt_path` passed to fit.
- `test_eval_always_passes_ckpt_path` — `ckpt_path` always passed to test.
- `test_no_load_creates_fresh_model` — no `--load`, normal constructor, `ckpt_path=None`.

**Other test files**
- Removed the now-unused `restore_opt = False` from `ModelArgs` in `test_training.py`,
`test_clamping.py`, `test_plotting.py`, and `test_datasets.py`.

**Behaviour Table**

| CLI flags | Model instantiation | `trainer.fit` `ckpt_path` |
|---|---|---|
| (none) | `ModelClass(args, ...)` | `None` |
| `--load PATH` | `load_from_checkpoint(PATH, ...)` | `None` |
| `--load PATH --restore_opt` | `ModelClass(args, ...)` | `PATH` |
| `--eval test --load PATH` | `load_from_checkpoint(PATH, ...)` | N/A (`trainer.test`) |

### Issue Link

Closes #131 

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [ ] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by Mani212005** [comment]
@observingClouds @joeloskarsson Kindly review this code. If you think something should be implemented differently. Please let me know

File: tests/test_checkpoint_restore.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    that a checkpoint path is required for restoring training state.
+    """
+    # Standard library
+    from argparse import Namespace
+
+    args = Namespace(load=None, restore_opt=True)
+    with pytest.raises(AssertionError, match="not loading a checkpoint"):
+        assert args.load or not args.restore_opt, (
+            "Can not restore training state " "when not loading a checkpoint"
+        )
```
Comment (joeloskarsson) [inline tests/test_checkpoint_restore.py]: What does this test other than the namespace that was just created here? Does not seem to actually test any code from the library.

File: neural_lam/train_model.py [hunk 2]
```diff
@@ -82,7 +82,8 @@ def main(input_args=None):
     parser.add_argument(
         "--restore_opt",
         action="store_true",
-        help="If optimizer state should be restored with model",
+        help="If full training state should be restored with model "
+        "(default: false)",
```
Comment (joeloskarsson) [inline neural_lam/train_model.py]: Should we now also change the name of `--restore_opt` to something more descriptive? That makes it clear that it is not just about the optimizer. Maybe `--load_training_state`? I am open to other suggestions as well.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Thanks! I think this looks really great. Had a couple small thoughts that I commented on here

File: tests/test_checkpoint_restore.py (hunk 1 above)
Comment (Mani212005) [reply tests/test_checkpoint_restore.py]: you're right, that test didn't actually test any library code. I've updated it to properly test the assertion in the [main() function. Sorry about this. I have fixed the test: 
Now, test: 
-> Calls the actual `[main()] ` function from the library with  `--load_training_state`  but no ` --load `
->  Bypasses the `@logger.catch` decorator using `main.__wrapped__` to allow the `AssertionError` to propagate
-> Verifies that the assertion in `train_model.py `(line 250-252) is actually triggered

File: neural_lam/train_model.py (hunk 2 above)
Comment (Mani212005) [reply neural_lam/train_model.py]: I like this is suggestion to rename to `--load_training_state`  since it's more descriptive .  I have gone by it and implemented this on the code

File: tests/test_checkpoint_restore.py (hunk 1 above)
Comment (joeloskarsson) [reply tests/test_checkpoint_restore.py]: That's nice :)

File: neural_lam/train_model.py [hunk 3]
```diff
@@ -80,9 +80,10 @@ def main(input_args=None):
         help="Path to load model parameters from",
     )
     parser.add_argument(
-        "--restore_opt",
+        "--load_training_state",
         action="store_true",
-        help="If optimizer state should be restored with model",
+        help="If full training state should be restored with model "
+        "(default: false)",
```
Comment (joeloskarsson) [inline neural_lam/train_model.py]:
```suggestion
        help="If full training state should be restored with model"
```
As we have `ArgumentDefaultsHelpFormatter` I don't think this is needed.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Sorry for taking so long to look at this again, it drowned below a lot of other things in my inbox :/ 
Anyway, I think this looks pretty good, just had one documentation suggestion. And unfortunately (since I took so long here) I think the branch needs to be updated with all the recent changes to main, before we can merge. In particular make sure that the changelog entry ends up in the right place with all the additions there.

**Comment by Mani212005** [comment]
No worries :)
I would do the required changes and push the updates

File: neural_lam/train_model.py [hunk 4]
```diff
@@ -251,6 +251,9 @@ def main(input_args=None):
         ),
     )
     args = parser.parse_args(input_args)
+    # In tests parse_args may be patched with MagicMock. Normalize the flag
+    # to avoid truthy MagicMock attributes changing control flow.
+    load_training_state = bool(vars(args).get("load_training_state", False))
```
Comment (joeloskarsson) [inline neural_lam/train_model.py]: I don't fully follow this, and this seems quite hard to read now. It would be better to not adapt the code to the testing framework. If testing does not work without this then we need to reconsider the testing approach, not change the actual codebase IMO.

Comment (joeloskarsson) [reply neural_lam/train_model.py]: Will the refactoring in #208 make this easier? It should relieve some need for mocking argparse, and it will be merged before this PR.

**Comment by Mani212005** [comment]
Sorry for taking such a long time to revert.  I’ve been busy. would update the required changes on weekend. 
Thanks!

**Comment by joeloskarsson** [comment]
#208 is merged now, which means 1) sadly some conflicts and this needing to be adapted, but also 2) maybe https://github.com/mllam/neural-lam/pull/240#discussion_r3113208921 will be easier to sort out?

**Review by sadamov — APPROVED** [review]
I rebased on main which made this PR much lighter. @Mani212005 and @joeloskarsson please have a look if this is still in line with your original intent.

File: CHANGELOG.md [hunk 5]
```diff
@@ -70,6 +70,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Fix `WeatherDataset.__len__` off-by-one in analysis mode (was undercounting by 1 sample), include `num_past_forcing_steps` in the forecast-mode minimum-horizon check, validate forcing-side forecast horizon when forcing is present, and use `min(n_state, n_forcing)` when both are present in analysis mode; raise `IndexError` for out-of-range indices in `WeatherDataset.__getitem__` (with Python-style negative indexing support) [\#312](https://github.com/mllam/neural-lam/pull/312) @kshirajahere
 
+- Fix `--load` to truly restore weights only by default: instead of overriding the optimizer state inside `on_load_checkpoint` (which left epoch / scheduler / callbacks inherited from the checkpoint), `train_model.py` now reconstructs the model via `ForecasterModule.load_from_checkpoint` and calls `trainer.fit(model)` with no `ckpt_path`, so all non-weight training state starts fresh. Also rename the CLI flag from `--restore_opt` to `--load_training_state` to reflect that it controls more than just the optimizer (epoch / scheduler / callbacks too), and assert it requires `--load` [\#240](https://github.com/mllam/neural-lam/pull/240) @Mani212005
+
```
Comment (joeloskarsson) [inline CHANGELOG.md:74]: Please simplify this changelog entry. The point is to get an idea of the change at a glance, not for this to be a full description of the change done (that's what the PR description is for 😄).

**Review by joeloskarsson — APPROVED** [review]
Great! Yes, I am also happy with this now. Just added a small comment about simplifying the changelog, which can be done before merging this (does not affect code).
